src/stockflow/StockFlowUnits.jl

```diff
@@ -0,0 +1,355 @@
+const StockAndFlowU = StockAndFlowUUntyped{Symbol,Symbol,Int8,Float64}
+
+
+
+
+nu(p::StockAndFlow0U) = nparts(p, :U)
+ndu(p::StockAndFlow0U) = nparts(p, :DU)
+nlu(p::Union{StockAndFlowU, StockAndFlow0U}) = nparts(p, :LU)
+
+
+add_unit!(p::StockAndFlow0U;kw...) = add_part!(p,:U;kw...)
+add_units!(p::StockAndFlow0U,n;kw...) = add_parts!(p,:U,n;kw...)
+
+add_dunit!(p::StockAndFlow0U;kw...) = add_part!(p,:DU;kw...)
+add_dunits!(p::StockAndFlow0U,n;kw...) = add_parts!(p,:DU,n;kw...)
+
+add_UDUlink!(p::StockAndFlow0U,u,du;kw...) = add_part!(p,:LU;luu=u,ludu=du,kw...)
+add_UDUlinks!(p::StockAndFlow0U,n,u,du;kw...) = add_parts!(p,:LU,n;luu=u,ludu=du, kw...)
+
+uname(p::StockAndFlow0U,u) = subpart(p,u,:uname) 
+unames(p::StockAndFlow0U) = subpart(p,:uname) 
+
+set_unames!(p::StockAndFlow0U, names) = set_subpart!(p, :uname, names)
+set_dunames!(p::StockAndFlow0U, names) = set_subpart!(p, :duname, names)
+set_exps!(p::StockAndFlow0U, exps) = set_subpart!(p, :exp, exps)
+
+
+set_sdu!(p::StockAndFlow0U, du) = set_subpart!(p, :sdu, du)
+
+
+
+add_unit!(p::StockAndFlowU;kw...) = add_part!(p,:U;kw...)
+add_units!(p::StockAndFlowU,n;kw...) = add_parts!(p,:U,n;kw...)
+
+add_dunit!(p::StockAndFlowU;kw...) = add_part!(p,:DU;kw...)
+add_dunits!(p::StockAndFlowU,n;kw...) = add_parts!(p,:DU,n;kw...)
+
+add_UDUlink!(p::StockAndFlowU,u,du;kw...) = add_part!(p,:LU;luu=u,ludu=du,kw...)
+add_UDUlinks!(p::StockAndFlowU,n,u,du;kw...) = add_parts!(p,:LU,n;luu=u,ludu=du, kw...)
+
+uname(p::StockAndFlowU,u) = subpart(p,u,:uname) 
+unames(p::StockAndFlowU) = subpart(p,:uname) 
+
+duname(p::Union{StockAndFlowU, StockAndFlow0U}, u) = subpart(p,u, :duname) 
+dunames(p::Union{StockAndFlowU, StockAndFlow0U}) = subpart(p,:duname) 
+
+
+set_unames!(p::StockAndFlowU, names) = set_subpart!(p, :uname, names)
+set_dunames!(p::StockAndFlowU, names) = set_subpart!(p, :duname, names)
+set_exps!(p::StockAndFlowU, exps) = set_subpart!(p, :exp, exps)
+get_exps(p::StockAndFlowU) = subpart(p, :exp)
+
+
+set_sdu!(p::StockAndFlowU, du) = set_subpart!(p, :sdu, du)
+set_pdu!(p::StockAndFlowU, du) = set_subpart!(p, :pdu, du)
+
+
+get_sdus(p::Union{StockAndFlowU, StockAndFlow0U}) = subpart(p, :sdu)
+get_sdu(p::Union{StockAndFlowU, StockAndFlow0U}, s) = subpart(p, :sdu)[s]
+
+get_pdus(p::StockAndFlowU) = subpart(p, :pdu)
+get_pdu(p::StockAndFlowU, param) = subpart(p, :pdu)[param]
+
+# get_sdu(p::StockAndFlowU, du) = subpart(p, :pdu, )
+
+
+lunames(p::Union{StockAndFlowU, StockAndFlow0U}) = begin
+    luu = map(x->subpart(p,x,:luu),collect(1:nlu(p)))
+    ludu = map(x->subpart(p,x,:ludu),collect(1:nlu(p)))
+    unit_names = map(x->uname(p,x),luu)
+    dunit_names = map(x->duname(p,x),ludu)
+    plu = collect(zip(unit_names, dunit_names))
+end
+
+
+const OpenStockAndFlowUOb, OpenStockAndFlowU = OpenACSetTypes(StockAndFlowUUntyped,StockAndFlowUntyped0U)
+
+
+
+leg(a::StockAndFlow0U, x::StockAndFlowU) = begin
+    if ns(a)>0 # if have stocks
```

Comment:
Would it make sense to DRY this up a bit? E.g., for each function in [ns, nsv, nls, nu, ndu, nlu], do the thing and collect into result?

Reply:
Yeah that'd make sense.  There are a few ways it could be compressed.  This was a result of me just copying what was done for StockAndFlowF

Reply:
Really, ntcomponent should just return an Int Vector regardless, and this check should be unnecessary